### ISSUE TITLE
Composer executable lookup

### ISSUE THREAD

**Issue by maff** [pimcore/pimcore#1362]
Make sure executables are correctly loaded from `$PATH`:

```php
$binary = 'composer';

var_dump([
    'path' => getenv('PATH'),
    'sys'  => exec('which ' . $binary),
    'tool' => Tool\Console::getExecutable($binary)
]);
```

```
array (size=3)
  'path' => string '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' (length=60)
  'sys' => string '/usr/local/bin/composer' (length=23)
  'tool' => boolean false
```

**Comment by tfarmerjms** [comment]
This is still an issue. Fresh install of Pimcore v5.10 and the System Requirements check complained with the red exclamation that it could not find composer. Composer is installed in /usr/bin/composer. If I su to the apache user and run it from the webroot, it is able to find composer. Only Pimcore cannot find it.

**Comment by IgorBenko** [comment]
@brusch @maff I was thinking about a fix for this issue, but didn't find a real solution yet.
The problem isn't that the composer can't be found, but rather the php can't be found. 
Composer has the following shebang:
#!/usr/bin/env php
The /usr/bin/env causes the problem. The environment often doesn't have the PATH defined when called from shell_exec, that's why you get "Permission denied" when trying to call '/usr/bin/env'.
The entered path in the Pimcore settings doesn't help in this case of course.

So the server must be either configured to have a PATH - maybe that's a check that you could do in the Console::getExecutable() method. Or composer could be executed with the php path prepended in the front of the command - this is less than optimal and than we would also need to figure out beforehand where the php executable is.

I think it should be also possible to just define the PATH before executing the command by doing this:
PATH=THE_PATH_FROM_PIMCORE_SETTINGS composer

If you find any of the suggestions above good, I would be glad to do a pull request.

**Comment by dpfaffenbauer** [comment]
@IgorBenko in that case: you can do following:

```
parameters
    pimcore_executable_composer: php composer.phar
```

So composer gets executed the correct way.

**Comment by IgorBenko** [comment]
@dpfaffenbauer All good ideas, thanks! I actually solved it by reconfiguring PHP-FPM. But this still doesn't solve the initial cryptic error message of the composer not being installed. Would like to see this solved at the core level. With a correct PATH being set in the Pimcore Settings, I wouldn't expect the Composer to still not be found - it guides user in the wrong direction and many users seem to have this issue.

**Comment by israasalameh95** [comment]
for which version this fix ??
I am using pimcore v4.6 and I encounter this issue , even I have the composer configured globally but the pimcore system check found that its not installed !! , btw: I am using windows machine

**Comment by Cruiser13** [comment]
Windows is not officially supported. If composer is properly setup on Windows Server with IIS, we simply used Di in Pimcore 4 to return composer directly. Worked fine. If you don't want to use DI you can simply change getExecutable() function in pimcore/lib/Pimcore/Tool/Console.php to return composer directly as string if it's searching for composer.

**Comment by israasalameh95** [comment]
actually I am using Xampp on windows machine

**Comment by primesoftnz** [comment]
I have the same problem as tfarmer00

**Comment by capitalfuse** [comment]
\website\config\di.example.php ---> di.php
and the add the following contents;

> <?php
> 
> // Pimcore is using PHP-DI, see http://php-di.org/doc/
> 
> return [
> 	"pimcore.executable.composer" => "/usr/local/bin/composer",
> ];

**Comment by primesoftnz** [comment]
Thank you

**Comment by brusch** [comment]
This topic should be already solved by abbe44d414dbe262ad24afc3fffae16406472ea0 .
